cmd/shell.go

```diff
@@ -336,16 +335,36 @@ func RunForever(server *rpc.AmqpRPCServer) {
 }
 
 // RunUntilSignaled starts the server and run until we get something on closeChan
-func RunUntilSignaled(logger *blog.AuditLogger, server *rpc.AmqpRPCServer, closeChan chan *amqp.Error) {
-	server.Start()
+func RunUntilSignaled(logger *blog.AuditLogger, server *rpc.AmqpRPCServer, closeChan chan *amqp.Error) error {
+	finishedProcessing, err := server.Start()
+	if err != nil {
+		return err
+	}
+	stopWatching, err := server.HandleInterrupts()
+	if err != nil {
+		return err
+	}
 	fmt.Fprintf(os.Stderr, "Server running...\n")
 
 	// Block until channel closes
-	err := <-closeChan
-
-	logger.Warning(fmt.Sprintf("AMQP Channel closed, will reconnect in 5 seconds: [%s]", err))
-	time.Sleep(time.Second * 5)
-	logger.Warning("Reconnecting to AMQP...")
+	for {
+		finished := true
+		select {
+		case err := <-closeChan:
+			logger.Warning(fmt.Sprintf("AMQP Channel closed, will reconnect in 5 seconds: [%s]", err))
```

Comment:
I think this reconnect logic for AMQP should be moved inside amqp-rpc.go.

Also, it appears that in this loop finished never becomes false, and can be removed. But I suspect this loop will look different once you move it into amqp-rpc.go.